### PR TITLE
[SPARK-46366][SQL] Use WITH expression in BETWEEN to avoid duplicate expressions

### PR DESCRIPTION
**Pull request by dbatomic**
### What changes were proposed in this pull request?

Prior to this change `e BETWEEN lower AND upper` expression used to be transformed into `lower <= e && e <= upper`. This means that `e` would be evaluated twice which is problematic from both correctness and performance perspectives.

Suggested fix is to use `WITH` expression that was introduced with [this](https://github.com/apache/spark/commit/01c294b05f3a9b7bd87cda0ee8b0160f5f58bb24) change.

### Why are the changes needed?

Current implementation is not correct for non deterministic expressions, since two calls might return different results.

### Does this PR introduce _any_ user-facing change?

With this change generated plan for BETWEEN statement will be different. An example of generated plan is provided in tests.

### How was this patch tested?

Existing tests plus new test in PlanGenerationTestSuite.

### Was this patch authored or co-authored using generative AI tooling?

Yes.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/betweenExpression.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  usage = "BETWEEN(projection, lower, upper) - Returns true if `projection` > `lower` and `projection` <  `upper`.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(0.5, 0, 1);
+        TRUE
+  """,
+  since = "2.0.0",
+  group = "conditional_funcs")
+case class BetweenExpr(proj: Expression, lower: Expression,
+                       upper: Expression, replacement: Expression)
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/betweenExpression.scala:31]: Please, adjust indentations according to https://github.com/databricks/scala-style-guide?tab=readme-ov-file#indent

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/betweenExpression.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "BETWEEN(projection, lower, upper) - Returns true if `projection` > `lower` and `projection` <  `upper`.",
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/betweenExpression.scala:22]: Could you describe arguments too in `arguments`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -3436,6 +3437,21 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
     }
   }
 
+  /**
+   * Transform UnresolvedBetweenExpression into a [BetweenExpr].
+   */
+  object ResolveBetweenExpression extends Rule[LogicalPlan] {
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3443]: We should avoid trying to add new analyzer rules at this point. Could we accomplish this  by combining `UnresolvedBetweenExpression` and `ResolvedBetweeExpression` into just one `BetweenExpression` and make that inherit from `RuntimeReplaceable`? Then its `replacement` can be the `With` expression of interest.

image

Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3443]: This is what I tried initially. I tried to explain the problem in PR description, but I guess that I wasn't clear enough. Let me try to do this again:
Initially I tried with `BetweenExpr extends RuntimeReplaceable`. The problem that I hit was that `CommonExpressionRef` required dataType of `CommonExpressionDef` to be known for it to be created:
```scala
  def this(exprDef: CommonExpressionDef) = this(exprDef.id, exprDef.dataType, exprDef.nullable)
```
To get around this I would like to create replacement of BetweenExpr after I resolve it's Child expressions and hence after I have it's dataType. I hit a wall doing this so I did this thing with UnresolvedBetween->BetweenExpr (for which I will know the types)->With, which works, but I agree that it would be better if we could avoid this + get rid of extra rule.

I will give it another try. If you guys know a way to get around this I would appreciate some help :)

Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Ok, I hope that i got it right in the latest iteration. Please take a look.

Btw, thanks for the comment, this is definitely much cleaner.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 4]
```diff
@@ -775,6 +775,7 @@ object FunctionRegistry {
     expression[In]("in"),
     expression[Not]("not"),
     expression[Or]("or"),
+    expression[Between]("between"),
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:778]: please sort alphabetically (move to line 775)?

**Review by dtenedor — APPROVED** [review]
LGTM!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala [hunk 5]
```diff
@@ -96,17 +96,22 @@ object StreamingJoinHelper extends PredicateHelper with Logging {
...
-        case LessThanOrEqual(l, r) => getStateWatermarkSafely(l, r).map(_ - 1)
-        case GreaterThan(l, r) => getStateWatermarkSafely(r, l)
-        case GreaterThanOrEqual(l, r) => getStateWatermarkSafely(r, l).map(_ - 1)
-        case _ => None
-      }
-      if (stateWatermark.nonEmpty) {
-        logInfo(s"Condition $joinCondition generated watermark constraint = ${stateWatermark.get}")
-      }
-      stateWatermark
+      val stateWatermarks = (predicate match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala:99]: why do we change the streaming code?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 6]
```diff
@@ -479,10 +479,8 @@ abstract class TypeCoercionBase {
         findWiderCommonType(types) match {
           case Some(finalDataType) =>
             Coalesce(es.map(castIfNotSameType(_, finalDataType)))
-          case None =>
-            c
+          case None => c
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:482]: let's avoid unnecessary changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  """,
+  arguments = """
+    Arguments:
+      * projection - An expression that is being compared with lower and upper bound.
+      * lower - Lower bound of the between check.
+      * upper - Upper bound of the between check.
+  """,
+  since = "4.0.0",
+  group = "conditional_funcs")
+case class Between(proj: Expression, lower: Expression, upper: Expression, replacement: Expression)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:36]:
```suggestion
case class Between private(proj: Expression, lower: Expression, upper: Expression, replacement: Expression)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  extends RuntimeReplaceable with InheritAnalysisRules  {
+  def this(proj: Expression, lower: Expression, upper: Expression) = {
+    this(proj, lower, upper, {
+        val commonExpr = CommonExpressionDef(proj)
+        val ref = new CommonExpressionRef(commonExpr)
+        With(And(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:42]: to make the code a bit more clearer:
```
val replacement = And(GreaterThanOrEqual(ref, lower), LessThanOrEqual(ref, upper))
With(replacement, Seq(commonExpr))
```

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala [hunk 9]
```diff
@@ -516,6 +516,10 @@ class PlanGenerationTestSuite
     simple.where("a + id < 1000")
   }
 
+  test("between expr") {
+    simple.selectExpr("rand(123) BETWEEN 0.1 AND 0.2")
```
Comment (cloud-fan) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:520]: can we add this test to a golden file test as well? This one is only for Spark Connect, we need to test the classic Spark SQL too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala (hunk 5 above)
Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala:99]: I got failures in streaming tests. It turns out that they calculate upper/lower bounds based on `if x > y` conditions. With `Between` op we no longer have expressions they can use so I had to extend the logic to cover `Between` for this watermark resolution.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala:99]: I think we can keep the code structure unchanged, with one more patter match
```
case Between(p, lower, upper, _) =>
  getStateWatermarkSafely(lower, p).map(_ - 1).orElse(getStateWatermarkSafely(p, upper).map(_ - 1))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+@ExpressionDescription(
+  usage = "Usage: expr1 [NOT] BETWEEN expr2 AND expr3 - evaluate if `expr1` is [not] in between `expr2` and `expr3`",
+  examples = """
+    Examples:
+      > SELECT 0.5 _FUNC_ 0.1 AND 1.0;
+        true
+  """,
+  arguments = """
+    Arguments:
+      * projection - An expression that is being compared with lower and upper bound.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:30]: The parameter name `projection` is a bit weird. How about just `value` or `input`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+      * lower - Lower bound of the between check.
+      * upper - Upper bound of the between check.
+  """,
+  since = "4.0.0",
+  group = "conditional_funcs")
+case class Between private(proj: Expression, lower: Expression, upper: Expression, replacement: Expression)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:36]: ditto, `value` or `input`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  extends RuntimeReplaceable with InheritAnalysisRules  {
+  def this(proj: Expression, lower: Expression, upper: Expression) = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:38]: ditto

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala (hunk 9 above)
Comment (cloud-fan) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:520]: I mean this golden file: sql/core/src/test/resources/sql-tests/inputs/predicate-functions.sql

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "Usage: expr1 [NOT] BETWEEN expr2 AND expr3 - evaluate if `expr1` is [not] in between `expr2` and `expr3`",
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:22]: let's use the parameter names in the usage doc as well.

**Comment by cloud-fan** [comment]
let's not forget about this comment :)  https://github.com/apache/spark/pull/44299/files#r1436957948

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala (hunk 9 above)
Comment (cloud-fan) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:520]: ping @dbatomic , this is not addressed yet.

**Comment by cloud-fan** [comment]
thanks, merging to master!
